lib/loglimit/loglimit_test.go

```diff
@@ -0,0 +1,183 @@
+/*
+Copyright 2023 Gravitational, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+	http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+package loglimit
+
+import (
+	"testing"
+	"time"
+
+	"github.com/jonboulle/clockwork"
+	log "github.com/sirupsen/logrus"
+	logtest "github.com/sirupsen/logrus/hooks/test"
+	"github.com/stretchr/testify/require"
+)
+
+func TestLogLimiter(t *testing.T) {
+	t.Parallel()
+	testCases := []struct {
+		desc             string
+		logSubstrings    []string
+		logsFirstBatch   []string
+		logsSecondsBatch []string
+		logsAssert       func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string)
+	}{
+		{
+			desc: "logs that do not match log substrings are logged",
+			logSubstrings: []string{
+				"A",
+			},
+			logsFirstBatch: []string{
+				"B log 1",
+				"B log 2",
+				"C log 1",
+			},
+			logsSecondsBatch: []string{},
+			logsAssert: func(t *testing.T, loggedFirstBatch, loggedSecondBatch []string) {
+				require.Len(t, loggedFirstBatch, 3)
+				require.Len(t, loggedSecondBatch, 0)
+
+				require.Equal(t, loggedFirstBatch[0], "B log 1")
+				require.Equal(t, loggedFirstBatch[1], "B log 2")
+				require.Equal(t, loggedFirstBatch[2], "C log 1")
```

Comment:
1. Use rich assertions
2. Use `assert` instead of `require`, if possible, so we get a more complete error report

```suggestion
				assert.Equal(t, []string{"B log 1", "B log 2", "C log 1"}, loggedFirstBatch, "first batch elements mismatch")
				assert.Empty(t, loggedSecondBatch, "second batch not empty)
```

Same for others.

Reply:
Improved this in 27f19cc654c3f99c2c2113d6394a5780f5da70c0. Still using `require` instead of `assert` as that's the recommendation in RFD 1.

Reply:
I think you are taking it too literally - I suspect the author forgot or was unaware about `assert` when they wrote that.

Despite our use of testify, I'd say this is applicable: https://github.com/golang/go/wiki/TestComments#keep-going.

Reply:
Ah, I see, thanks! The difference seems to be, not in the output, but in the fact that `assert` allows the test to keep going while `require` fails the test right away. Replaced in 12282ba392967c6aa9028985d9fac4d544d0e4c9.